File: 2019/day03-2.ts
type Coords = {
  x: number;
  y: number;
};
type Instruction = {
  direction: string;
  amount: number;
};
type Segment = {
  start: Coords;
  end: Coords;
};

function solution(content: string) {
  const lines = content.split("\n");
  const seg1 = parse(lines[0]);
  const seg2 = parse(lines[1]);
  const wire1 = untangle(seg1);
  const wire2 = untangle(seg2);
  return seg1
    .map((s1) => seg2.filter((s2) => intersect(s1, s2)).map((s2) => [s1, s2]))
    .flat()
    .map(([s1, s2]) =>
      unfold(s1).find((coords) =>
        (coords.x !== 0 || coords.y !== 0) &&
        unfold(s2).find((item) => item.x === coords.x && item.y === coords.y)
          ? coords
          : null
      )
    )
    .filter((item) => item)
    .map((coords) => length(coords!, wire1) + length(coords!, wire2))
    .toSorted()
    .at(0);
}

function length(coords: Coords, wire: Coords[]) {
  return wire.findIndex(
    (value) => value.x === coords.x && value.y === coords.y
  );
}

function untangle(wire: Segment[]): Coords[] {
  return wire.map((segment) => unfold(segment)).flat();
}

function unfold(segment: Segment): Coords[] {
  const {
    start: { x: startX, y: startY },
    end: { x: endX, y: endY },
  } = segment;
  if (isVertical(segment)) {
    const sign = Math.sign(endY - startY);
    return Array.from({ length: Math.abs(startY - endY) }).map((_, index) => ({
      x: startX,
      y: startY + index * sign,
    }));
  }
  const sign = Math.sign(endX - startX);
  return Array.from({ length: Math.abs(startX - endX) }).map((_, index) => ({
    x: startX + index * sign,
    y: startY,
  }));
}

function intersect(s1: Segment, s2: Segment) {
  const {
    start: { x: start1x, y: start1y },
    end: { x: end1x, y: end1y },
  } = s1;
  const {
    start: { x: start2x, y: start2y },
    end: { x: end2x, y: end2y },
  } = s2;
  if (isVertical(s1) && isVertical(s2))
    return (
      start1y === start2y &&
      (between(start1y, start2y, end2y) ||
        between(end1y, start2y, end2y) ||
        between(start2y, start1y, end1y) ||
        between(end2y, start1y, end1y))
    );
  if (!isVertical(s1) && isVertical(s2))
    return between(start1y, start2y, end2y) && between(start2x, start1x, end1x);
  if (isVertical(s1) && !isVertical(s2))
    return between(start1x, start2x, end2x) && between(start2y, start1y, end1y);
  return (
    start1x === start2x &&
    (between(start1x, start2x, end2x) ||
      between(end1x, start2x, end2x) ||
      between(start2x, start1x, end1x) ||
      between(end2x, start1x, end1x))
  );
}

function isVertical({ start: { x: startX }, end: { x: endX } }: Segment) {
  return startX === endX;
}

function between(x: number, start: number, end: number) {
  return start < end ? start <= x && x <= end : end <= x && x <= end;
}

function parse(line: string): Segment[] {
  return line
    .split(",")
    .map((item) => parseInstruction(item))
    .reduce((acc, instr) => {
      const last = acc.at(-1);
      if (last) {
        return [
          ...acc,
          { start: last.end, end: move(last.end, instr) },
        ] as Segment[];
      }
      return [{ start: { x: 0, y: 0 }, end: move({ x: 0, y: 0 }, instr) }];
    }, [] as Segment[]);
}

function move({ x, y }: Coords, { direction, amount }: Instruction): Coords {
  if (direction === "R") return { x: x + amount, y };
  if (direction === "L") return { x: x - amount, y };
  if (direction === "U") return { x, y: y + amount };
  if (direction === "D") return { x, y: y - amount };
  return { x, y };
}

function parseInstruction(word: string): Instruction {
  return {
    amount: Number(word.slice(1)),
    direction: word[0],
  };
}
Deno.readTextFile("day03.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2019/day03_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import Any

type Coords = tuple[int, int]
type Instruction = tuple[str, int]
type Segment = tuple[Coords, Coords]


def solution(content: str) -> Any:
    lines = content.split("\n")
    w1_seg = parse(lines[0])
    w2_seg = parse(lines[1])
    wire1 = untangle(w1_seg)
    wire2 = untangle(w2_seg)

    return sorted(
        length(coords, wire1) + length(coords, wire2)
        for s1, s2 in [(s1, s2) for s1 in w1_seg for s2 in w2_seg if intersect(s1, s2)]
        for coords in unfold(s1)
        if coords in unfold(s2) and coords != (0, 0)
    )[0]


def length(coords: Coords, wire: list[Coords]):
    return wire.index(coords)


def untangle(wire: list[Segment]):
    return [coords for segment in wire for coords in unfold(segment)]


def unfold(segment: Segment):
    start, end = segment
    if is_vertical(segment):
        return [
            (start[0], start[1] + y * (1 if end[1] > start[1] else -1))
            for y in range(abs(start[1] - end[1]))
        ]
    return [
        (start[0] + x * (1 if end[0] > start[0] else -1), start[1])
        for x in range(abs(start[0] - end[0]))
    ]


def intersect(s1: Segment, s2: Segment):
    start1, end1 = s1
    start2, end2 = s2
    if is_vertical(s1) and is_vertical(s2):
        return start1[0] == start2[0] and (
            between(start1[1], start2[1], end2[1])
            or between(end1[1], start2[1], end2[1])
            or between(start2[1], start1[1], end1[1])
            or between(end2[1], start1[1], end1[1])
        )
    if not is_vertical(s1) and is_vertical(s2):
        return between(start1[1], start2[1], end2[1]) and between(
            start2[0], start1[0], end1[0]
        )
    if is_vertical(s1) and not is_vertical(s2):
        return between(start1[0], start2[0], end2[0]) and between(
            start2[1], start1[1], end1[1]
        )
    if not is_vertical(s1) and not is_vertical(s2):
        return start1[1] == start2[1] and (
            (
                between(start1[0], start2[0], end2[0])
                or between(end1[0], start2[0], end2[0])
                or between(start2[0], start1[0], end1[0])
                or between(end2[0], start1[0], end1[0])
            )
        )
    return False


def is_vertical(seg: Segment):
    start, end = seg
    return start[0] == end[0]


def between(x: int, start: int, end: int):
    return (start <= x <= end) if start < end else (end <= x <= start)


def parse(line: str) -> list[Segment]:
    segments = []
    for instruction in [parse_instruction(word) for word in line.split(",")]:
        if segments:
            _, last = segments[-1]
            segments.append((last, move(last, instruction)))
        else:
            segments.append(((0, 0), move((0, 0), instruction)))
    return segments


def move(start: Coords, instr: Instruction):
    if instr[0] == "R":
        return (start[0] + instr[1], start[1])
    if instr[0] == "L":
        return (start[0] - instr[1], start[1])
    if instr[0] == "U":
        return (start[0], start[1] + instr[1])
    if instr[0] == "D":
        return (start[0], start[1] - instr[1])
    return start


def parse_instruction(word: str) -> Instruction:
    return (word[0], int(word[1:]))


def manhattan(c1: Coords, c2: Coords):
    return abs(c1[0] - c2[0]) + abs(c1[1] - c2[1])


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
